src/main/scala/Engine.scala

```diff
@@ -22,6 +35,92 @@ trait Engine:
   def find[B <: Identifiable](using
       tt: TypeTest[Behaviour, B]
   )(id: String): Option[B]
-// TODO: add tests before decommenting
-// extension (engine: Engine)
-//   def deltaTimeSeconds: Double = engine.deltaTimeNanos / Math.pow(10, 9)
+
+object Engine:
+  //gameObjects, numSteps and dtNanos are used just to test the Engine until other Interfaces are implemented
+  private class EngineImpl(
+      override val io: IO,
+      override val storage: Storage,
+      gameObjects: Iterable[Behaviour],
+      numSteps: Int,
+      dtNanos: Long
+  ) extends Engine:
+
+    override def loadScene(scene: Scene): Unit = ???
+
+    override def enable(gameObject: Behaviour): Unit = ???
+
+    override def disable(gameObject: Behaviour): Unit = ???
+
+    override def create(gameObject: Behaviour): Unit = ???
+
+    override def find[B <: Identifiable](using tt: TypeTest[Behaviour, B])(
+        id: String
+    ): Option[B] = ???
+
+    override def find[B <: Behaviour](using
+        tt: TypeTest[Behaviour, B]
+    )(): Iterable[B] = ???
+
+    override def deltaTimeNanos: Long = ???
+
+    override def destroy(gameObject: Behaviour): Unit = ???
+
+    private def enabledGameObjects =
+      gameObjects.filter(gameObject => gameObject.enabled)
+
+    private var currentStep = 0
+    private var shouldStop = false
+
+    override def run(): Unit =
+      gameObjects.foreach(gameObject => gameObject.onInit(this))
+
+      enabledGameObjects.foreach(gameObject => gameObject.onEnabled(this))
+
+      enabledGameObjects.foreach(gameObject => gameObject.onStart(this))
+
+      while !shouldStop && currentStep < numSteps do
+        enabledGameObjects.foreach(gameObject => gameObject.onEarlyUpdate(this))
+
+        enabledGameObjects.foreach(gameObject => gameObject.onUpdate(this))
+
+        enabledGameObjects.foreach(gameObject => gameObject.onLateUpdate(this))
+        currentStep = currentStep + 1
+
+      gameObjects.foreach(gameObject => gameObject.onDeinit(this))
+      currentStep = 0
+
+    override def stop(): Unit = shouldStop = true
+
+  var engine: Engine = null
```

Comment:
Se ho capito bene l'idea implementata qui più sotto è quella di avere un singleton per Engine, è corretto?
Siamo sicuri che aggiunga valore alla code-base? Quali sono i casi in cui avere questo singleton diventa importante?
I behaviour ricevono il riferimento attraverso i loro metodi, mentre negli altri componenti non abbiamo identificato dipendenze da Engine.